apps/epic-web/src/inngest/functions/stripe/checkout-completed.ts

```diff
@@ -1,16 +1,94 @@
 import {inngest} from 'inngest/inngest.server'
 import {STRIPE_CHECKOUT_COMPLETED_EVENT} from '@skillrecordings/inngest'
+import {sanityClient} from '@skillrecordings/skill-lesson/utils/sanity-client'
+import * as yaml from 'js-yaml'
+import {prisma} from '@skillrecordings/database'
+import {z} from 'zod'
+import {Redis} from '@upstash/redis'
+
+const redis = new Redis({
+  url: process.env.UPSTASH_REDIS_REST_URL!,
+  token: process.env.UPSTASH_REDIS_REST_TOKEN!,
+})
+
+const filterSchema = z.object({
+  ppp: z.boolean().optional(),
+  quantity: z.number().optional(),
+})
+
+function epochToISOString(epoch: number): string {
+  const date = new Date(epoch)
+  return date.toISOString()
+}
 
 export const stripeCheckoutCompleted = inngest.createFunction(
   {id: `stripe-checkout-completed`, name: 'Stripe Checkout Completed'},
   {event: STRIPE_CHECKOUT_COMPLETED_EVENT},
   async ({event, step}) => {
-    const {quantity} = event.data
+    const {quantity, created, purchaseId} = event.data
 
     const teamPurchase = quantity > 1
 
+    const purchase = await step.run('load purchase', async () => {
+      return prisma.purchase.findUnique({
+        where: {
+          id: purchaseId,
+        },
+      })
+    })
+
+    if (!purchase) {
+      return 'invalid purchase id'
+    }
+
     if (teamPurchase) {
     } else {
+      const date = epochToISOString(created)
+      const availableBonuses = await step.run('get bonuses', async () => {
+        const query = `*[_type == "bonus" && expiresAt > $date && validFrom < $validFrom]{
+              title,
+              "slug": slug.current,
+              filter,
+              description,
+              expiresAt
+            }`
+        return sanityClient.fetch(query, {
+          date,
+          validFrom: new Date().toISOString(),
+        })
+      })
+
+      if (availableBonuses.length > 0) {
+        const applicableBonuses: string[] = []
+        for (const availableBonus of availableBonuses) {
+          try {
+            const filters = availableBonus.filter
+              ? filterSchema.parse(yaml.load(availableBonus.filter))
+              : {}
+            if (filters.quantity === 1 && purchase.bulkCouponId) {
+              break
+            }
+            if (filters.ppp === false && purchase.status !== 'Valid') {
+              break
+            }
+            applicableBonuses.push(availableBonus.slug)
+          } catch (e) {
+            console.error('invalid bonus', e)
+          }
+        }
```

Comment:
loading all the valid bonuses from sanity and then finding the bonuses that are available to the purchaser